core/src/main/java/io/grpc/internal/BinaryLogProvider.java

```diff
@@ -16,237 +16,33 @@
-import java.io.IOException;
-import java.io.InputStream;
-import java.util.ArrayList;
+import io.grpc.InternalServiceProviders;
+import io.grpc.InternalServiceProviders.PriorityAccessor;
 import java.util.Collections;
-import java.util.Comparator;
-import java.util.Iterator;
-import java.util.List;
-import java.util.ServiceConfigurationError;
-import java.util.ServiceLoader;
-import java.util.logging.Level;
-import java.util.logging.Logger;
 import javax.annotation.Nullable;
 
-public abstract class BinaryLogProvider implements Closeable {
-  private static final Logger logger = Logger.getLogger(BinaryLogProvider.class.getName());
-  private static final BinaryLogProvider PROVIDER = load(BinaryLogProvider.class.getClassLoader());
-  @VisibleForTesting
-  static final Marshaller<InputStream> IDENTITY_MARSHALLER = new IdentityMarshaller();
-
-  private final ClientInterceptor binaryLogShim = new BinaryLogShim();
-
-  /**
-   * Returns a {@code BinaryLogProvider}, or {@code null} if there is no provider.
-   */
-  @Nullable
-  public static BinaryLogProvider provider() {
-    return PROVIDER;
-  }
-
-  /**
-   * Wraps a channel to provide binary logging on {@link ClientCall}s as needed.
-   */
-  Channel wrapChannel(Channel channel) {
-    return ClientInterceptors.intercept(channel, binaryLogShim);
-  }
-
-  private static MethodDescriptor<InputStream, InputStream> toInputStreamMethod(
-      MethodDescriptor<?, ?> method) {
-    return method.toBuilder(IDENTITY_MARSHALLER, IDENTITY_MARSHALLER).build();
-  }
-
-  /**
-   * Wraps a {@link ServerMethodDefinition} such that it performs binary logging if needed.
-   */
-  final <ReqT, RespT> ServerMethodDefinition<?, ?> wrapMethodDefinition(
-      ServerMethodDefinition<ReqT, RespT> oMethodDef) {
-    ServerInterceptor binlogInterceptor =
-        getServerInterceptor(oMethodDef.getMethodDescriptor().getFullMethodName());
-    if (binlogInterceptor == null) {
-      return oMethodDef;
-    }
-    MethodDescriptor<InputStream, InputStream> binMethod =
-        BinaryLogProvider.toInputStreamMethod(oMethodDef.getMethodDescriptor());
-    ServerMethodDefinition<InputStream, InputStream> binDef = InternalServerInterceptors
-        .wrapMethod(oMethodDef, binMethod);
-    ServerCallHandler<InputStream, InputStream> binlogHandler = InternalServerInterceptors
-        .interceptCallHandler(binlogInterceptor, binDef.getServerCallHandler());
-    return ServerMethodDefinition.create(binMethod, binlogHandler);
-  }
-
-
-  @VisibleForTesting
-  static BinaryLogProvider load(ClassLoader classLoader) {
-    try {
-      return loadHelper(classLoader);
-    } catch (Throwable t) {
-      logger.log(
-          Level.SEVERE, "caught exception loading BinaryLogProvider, will disable binary log", t);
-      return null;
-    }
-  }
-
-  private static BinaryLogProvider loadHelper(ClassLoader classLoader) {
-    if (isAndroid()) {
-      return null;
-    }
+public abstract class BinaryLogProvider {
```

Comment:
This class is weird. I don't know why you're deviating from the standard Provider pattern.

I'm okay with stuff moving out of this class, but the providers should extend this class and things like priority()/isAvailable() should be here.

Reply:
Perhaps I misunderstood [this comment](https://github.com/grpc/grpc-java/pull/3917#discussion_r164028522):
>This was a mistake we made with ServerBuilder. Having the static factory of the class be on itself and be extensible made for confusing behavior. The factory and the Sink should be in separate classes.

Maybe if I renamed `BinaryLog` to `BinaryLogProvider` to follow the pattern, and renamed `BinaryLogProvider` to `BinaryLogInstance` things would be cleared up.